File: repo/react/markra/apps/web/src/runtime/web/ignore-rules.ts
import ignore from "ignore";
import type { WebDirectoryHandle } from "./types";

const markraIgnoreFileName = ".markraignore";
const builtInSkippedDirectoryNames = new Set([
  ".codex",
  ".git",
  ".markra-sync",
  ".obsidian",
  "build",
  "dist",
  "node_modules",
  "target"
]);

export type MarkdownIgnoreRules = {
  ignores: (relativePath: string, isDirectory: boolean) => boolean;
};

function normalizeRelativePath(path: string) {
  return path.replace(/\\+/gu, "/").replace(/^\.\/|^\/|\/$/gu, "");
}

async function readMarkraIgnore(root: WebDirectoryHandle) {
  try {
    const handle = await root.getFileHandle?.(markraIgnoreFileName);
    return handle ? await (await handle.getFile()).text() : "";
  } catch {
    // Missing or unreadable control files must not block opening a workspace.
    return "";
  }
}

export async function loadMarkdownIgnoreRules(
  root: WebDirectoryHandle,
  globalIgnoreRules = ""
): Promise<MarkdownIgnoreRules> {
  const matcher = ignore({ ignorecase: false });
  try {
    matcher.add(globalIgnoreRules);
  } catch {
    // Keep any valid rules already parsed and continue with workspace rules.
  }
  try {
    // Workspace rules are appended later so they can override global defaults.
    matcher.add(await readMarkraIgnore(root));
  } catch {
    // Invalid workspace rules fail open so users can still access and repair them.
  }

  return {
    ignores(relativePath, isDirectory) {
      const normalizedPath = normalizeRelativePath(relativePath);
      if (normalizedPath === markraIgnoreFileName) return true;

      const parts = normalizedPath.split("/").filter(Boolean);
      const directoryParts = isDirectory ? parts : parts.slice(0, -1);

      // Built-in exclusions stay authoritative so user negations cannot restore
      // generated or tool-heavy subtrees that Markra deliberately avoids scanning.
      if (directoryParts.some((part) => builtInSkippedDirectoryNames.has(part))) return true;

      return normalizedPath.length > 0 && matcher.ignores(isDirectory ? `${normalizedPath}/` : normalizedPath);
    }
  };
}

File: repo/react/markra/apps/desktop/src-tauri/src/markdown_files/ignore_rules.rs
use std::ffi::OsStr;
use std::path::{Path, PathBuf};

use ignore::gitignore::{Gitignore, GitignoreBuilder};

pub(crate) const MARKRA_IGNORE_FILE_NAME: &str = ".markraignore";

fn is_builtin_ignored_directory_name(name: &OsStr) -> bool {
    name.to_str().is_some_and(|name| {
        matches!(
            name,
            ".codex"
                | ".git"
                | ".markra-sync"
                | ".obsidian"
                | "build"
                | "dist"
                | "node_modules"
                | "target"
        )
    })
}

pub(crate) struct MarkdownIgnoreRules {
    global_rules: String,
    include_workspace_rules: bool,
    root: PathBuf,
    matcher: Gitignore,
}

impl MarkdownIgnoreRules {
    pub(crate) fn for_root(root: &Path, global_rules: Option<&str>) -> Self {
        Self::build(root, global_rules, true)
    }

    pub(crate) fn built_in_only(root: &Path) -> Self {
        Self::build(root, None, false)
    }

    fn build(root: &Path, global_rules: Option<&str>, include_workspace_rules: bool) -> Self {
        let global_rules = global_rules.unwrap_or_default().to_string();
        let mut builder = GitignoreBuilder::new(root);

        // Parse line-by-line so one invalid global pattern cannot discard valid rules.
        for line in global_rules.lines() {
            let _ = builder.add_line(None, line);
        }
        if include_workspace_rules {
            // Workspace rules are added last so their negations can override global defaults.
            // Partial file errors are intentionally ignored to keep the workspace repairable.
            let _ = builder.add(root.join(MARKRA_IGNORE_FILE_NAME));
        }
        let matcher = builder.build().unwrap_or_else(|_| Gitignore::empty());

        Self {
            global_rules,
            include_workspace_rules,
            root: root.to_path_buf(),
            matcher,
        }
    }

    pub(crate) fn reload(&mut self) {
        let root = self.root.clone();
        let global_rules = self.global_rules.clone();
        *self = Self::build(&root, Some(&global_rules), self.include_workspace_rules);
    }

    pub(crate) fn ignores(&self, path: &Path, is_directory: bool) -> bool {
        let Ok(relative_path) = path.strip_prefix(&self.root) else {
            return false;
        };

        if self.is_control_file(path) {
            return true;
        }

        let directory_path = if is_directory {
            relative_path
        } else {
            relative_path.parent().unwrap_or_else(|| Path::new(""))
        };

        // Built-in exclusions protect workspace performance and remain authoritative
        // even when a user rule attempts to negate one of them.
        if directory_path
            .components()
            .any(|component| is_builtin_ignored_directory_name(component.as_os_str()))
        {
            return true;
        }

        self.matcher
            .matched_path_or_any_parents(path, is_directory)
            .is_ignore()
    }

    pub(crate) fn is_control_file(&self, path: &Path) -> bool {
        path.parent() == Some(self.root.as_path())
            && path.file_name() == Some(OsStr::new(MARKRA_IGNORE_FILE_NAME))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    fn test_root(name: &str) -> PathBuf {
        std::env::temp_dir().join(format!(
            "markra-ignore-rules-{name}-{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .expect("system clock should be after epoch")
                .as_nanos()
        ))
    }

    #[test]
    fn applies_global_rules_before_workspace_rules() {
        let root = test_root("precedence");
        fs::create_dir_all(&root).expect("test root should be created");
        fs::write(root.join(MARKRA_IGNORE_FILE_NAME), "!keep.md\n")
            .expect("workspace rules should be written");

        let rules = MarkdownIgnoreRules::for_root(&root, Some("*.md\n"));

        assert!(!rules.ignores(&root.join("keep.md"), false));
        assert!(rules.ignores(&root.join("drop.md"), false));

        fs::remove_dir_all(root).expect("test root should be removed");
    }

    #[test]
    fn built_in_directories_remain_authoritative() {
        let root = test_root("builtins");
        let rules =
            MarkdownIgnoreRules::for_root(&root, Some("!node_modules/\n!node_modules/readme.md\n"));

        assert!(rules.ignores(&root.join("node_modules/readme.md"), false));
    }

    #[test]
    fn built_in_only_rules_keep_user_ignored_markdown_visible() {
        let root = test_root("built-in-only");
        fs::create_dir_all(root.join("drafts")).expect("draft folder should be created");
        fs::create_dir_all(root.join(".git")).expect("git folder should be created");
        fs::write(root.join(".markraignore"), "drafts/\n").expect("ignore file should be written");

        let rules = MarkdownIgnoreRules::built_in_only(&root);

        assert!(!rules.ignores(&root.join("drafts/hidden.md"), false));
        assert!(rules.ignores(&root.join(".git/readme.md"), false));
        fs::remove_dir_all(root).expect("test root should be removed");
    }

    #[test]
    fn matches_ignore_rules_case_sensitively() {
        let root = test_root("case-sensitive");
        let rules = MarkdownIgnoreRules::for_root(&root, Some("drafts/\n"));

        assert!(rules.ignores(&root.join("drafts/note.md"), false));
        assert!(!rules.ignores(&root.join("Drafts/note.md"), false));
    }
}
